### ISSUE TITLE
Open edX: course breadcrumbs link to "legacy" UI

### ISSUE THREAD

**Issue by pdpinch** [mitodl/mitxonline#162]
It appears that the breadcrumb links in frontend-app-learning don't follow the path-based routing setting (i.e., PUBLIC_PATH) 

### Steps to Reproduce

1. Navigate to any courseware page, for example  https://courses-qa.mitxonline.mit.edu/learn/course/course-v1:MITx+JPAL102x+3T2021/block-v1:MITx+JPAL102x+3T2021+type@sequential+block@ed96b6319d8a4e97b9d8ad72a0306106/block-v1:MITx+JPAL102x+3T2021+type@vertical+block@e397da6d190140f8a0d24059ce0898e6
2. Click on a link in the breadcrumbs, for example "Entrance Survey"

### Expected Behavior

The UX should remain the same, and the link should take you to https://courses-qa.mitxonline.mit.edu/learn/course/course-v1:MITx+JPAL102x+3T2021/block-v1:MITx+JPAL102x+3T2021+type@sequential+block@ed96b6319d8a4e97b9d8ad72a0306106/block-v1:MITx+JPAL102x+3T2021+type@vertical+block@e397da6d190140f8a0d24059ce0898e6

### Actual Behavior

- The link takes you to https://courses-qa.mitxonline.mit.edu/learn/course/course-v1:MITx+JPAL102x+3T2021/block-v1:MITx+JPAL102x+3T2021+type@sequential+block@ed96b6319d8a4e97b9d8ad72a0306106/block-v1:MITx+JPAL102x+3T2021+type@vertical+block@e397da6d190140f8a0d24059ce0898e6 white redirects to https://courses-qa.mitxonline.mit.edu/courses/course-v1:MITx+JPAL102x+3T2021/course/#block-v1:MITx+JPAL102x+3T2021+type@sequential+block@ed96b6319d8a4e97b9d8ad72a0306106
- The UX reverts to the "legacy" course experience

**Comment by pdpinch** [comment]
This seems to be a bug in frontend-app-learning that edX doesn't experience because they don't have it configured this way. 

We should fix this and upstream it.

**Comment by arslanashraf7** [comment]
**Update:**

As part of the initial update there are below findings that might need us to decide on how to proceed on this:

- Strangely, The breadcrumb redirect problem is not associated with `PUBLIC_PATH`, I've tested it with and without setting the path-based route but the behavior remains the same.
- This problem is specifically associated with the `User Role`. e.g. This problem of legacy redirect is only faced by the `staff users` either enrolled in a course or not and other general users do not face this problem. This can be tested on [our RC](https://courses-qa.mitxonline.mit.edu/) as well where accessing the courseware on MFE with a general user doesn't show this behavior but only the staff user.

Based on the above, There are some questions that need more digging:
- Is this behavior intentional for the `staff` users?
- There are a couple of PRs though regarding the kill-switches for the MFE visibility e.g. https://github.com/edx/edx-platform/pull/24772, https://github.com/edx/frontend-app-learning/pull/427, https://github.com/edx/frontend-app-learning/pull/404 but all this might need detailed debugging as staff role-related checks for MFE are at many places, and should we go for that?
- I think there might be a related waffle flag to keep the MFE experience for the staff role as well.

I'll post more details on this unless specified otherwise.

@pdpinch FYI.

**Comment by pdpinch** [comment]
When you say "staff" do you mean django staff, or course staff (or both)?

I'll reach out to some folks at edX to see if they can explain this behavior. But if it only affects staff, it seems like a low-priority issue.

**Comment by arslanashraf7** [comment]
> When you say "staff" do you mean django staff, or course staff (or both)?

Looks like it's happening with django staff.

**Comment by pdpinch** [comment]
Ok, after some effort I've been able to test on RC with a regular learner account without django or course staff privileges. 

On RC, it seems like the behavior is a little different for the learner view, but still not quite right. When I click the breadcrumb links, I stay in the "new learning experience" but no matter which breadcrumb I click on, it takes me to the course outline. I suspect this is because the URLs in the links are still being constructed with the legacy /courses/ path. 

Is there some way to get the MFE to honor the `PUBLIC_PATH` setting? I looked through various "legacy_frontend" toggles, but they all appear to default to off/false.

**Comment by arslanashraf7** [comment]
@pdpinch 

> On RC, it seems like the behavior is a little different for the learner view, but still not quite right. When I click the breadcrumb links, I stay in the "new learning experience" but no matter which breadcrumb I click on, it takes me to the course outline. I suspect this is because the URLs in the links are still being constructed with the legacy /courses/ path.

I was cross-matching the learning MFE behavior with one on [edX's courses](https://learning.edx.org/course/course-v1:edX+DemoX.1+2T2019/home#block-v1:edX+DemoX.1+2T2019+type@sequential+block@ae87afb9cd264a6bb31bbc1a155d96a0) which I enrolled in to see how edX has been behaving and there it also creates the breadcrumb URL with legacy `/courses/`. If you have an edX account and enroll in the Demo course you might be able to experience it too.


> Is there some way to get the MFE to honor the `PUBLIC_PATH` setting? I looked through various "legacy_frontend" toggles, but they all appear to default to off/false.

It's possible that there might be a way, but I'll have to look.

**Comment by arslanashraf7** [comment]
@pdpinch, Below are more details and a solution to the problem you mentioned in the last comment.

This might probably be the last update in order to resolve the issue overall:

**First Problem(Resolved per https://github.com/mitodl/mitxonline/issues/162#issuecomment-919351481):** It was our assumption that the custom `PUBLIC_PATH` for the Learning MFE was the reason that was preventing the users from keeping in the new learning experience and instead of that it was taking the users to the legacy experience, But that was not actually related to the `PUBLIC_PATH` but the User Role, specifically the `Django staff` role. Other users were unaffected by this behavior.

**Second Problem:** The second problem here was that even the redirect was not taking the `non-Django staff` users to the legacy but it was also not behaving as it should within the course line in the new MFE. It should automatically expand the last opened unit in the course line as it does in [edX](https://learning.edx.org/).

While figuring out what was happening in the second problem I figured that the `Start Course` button on the course Home Page[e.g. this page](https://courses-qa.mitxonline.mit.edu/learn/course/course-v1:PUCIT+CSF14+2014-2018/home) within the MFE was also not maintaining the state and the whenever the user would come back to the page the `Start Course` button would appear again.

Inside what was happening was that the value of `has_visited_course` from the API(`api/course_home/outline/<course_id>`) that MFE hits for course details because there was no progress being tracked for the user in edX upon opening a course outline link hence making the `Start Course` button appear always.
In addition to the above, we were also getting `"resume_block": false` for all the blocks inside a course so the learning MFE wouldn't know which block to expand inside the course outline.

**Long story short:** The course completion was not working at all for our courses.

**Solution:** We will need to add a waffle switch for `completion.enable_completion_tracking` in our [Admin waffle](https://courses-qa.mitxonline.mit.edu/admin/waffle/switch/), courtesy of [this post](https://discuss.openedx.org/t/publish-completion-never-seems-to-fire/628). This will enable the course completion feature and things would start to behave as expected.

**Comment by arslanashraf7** [comment]
@pdpinch based on the above comment what course of action do you suggest moving on with this ticket?

**Comment by pdpinch** [comment]
Ok, I turned on the switch on QA https://courses-qa.mitxonline.mit.edu/admin/waffle/switch/3/change/ -- how can I tell if it works?

Will this change the linking behavior in the MFE, or is that a separate issue?

**Comment by pdpinch** [comment]
Someone told me how to test it, and it works. I turned it on in production too.

**Comment by arslanashraf7** [comment]
> how can I tell if it works?
> Someone told me how to test it, and it works. I turned it on in production too.

Sorry I missed adding a response to that and if in your testing you managed to get an expanded course outline based on the last section/subsection you completed along with highlighted green tabs, that would probably be it to check that the completion tracking is working.

Along with the above, if we are not seeing any more legacy redirects and 404 upon clicking the breadcrumb link in the learning MFE, that should probably be it for closing this. And as its working now meeting the criteria of the ticket.
